segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -0,0 +1,385 @@
+     */
+    @Override
+    protected CompletableFuture<Void> doRun() {
+        loopFuture = Futures.loop(
+                this::canRun,
+                () -> delaySupplier.get()
+                        .thenRunAsync(() -> {
+                            log.info("{}: Iteration {} started.", traceObjectId, iterationId.get());
+                        }, executor)
+                        .thenComposeAsync(v -> deleteGarbage(true, config.getGarbageCollectionMaxConcurrency()), executor)
+                        .handleAsync((v, ex) -> {
+                            if (null != ex) {
+                                log.error("{}: Error during doRun.", traceObjectId, ex);
+                            }
+                            log.info("{}: Iteration {} ended.", traceObjectId, iterationId.getAndIncrement());
+                            return null;
+                        }, executor),
+                executor);
+        return loopFuture;
+    }
+
+    private boolean canRun() {
+        return isRunning() && getStopException() == null && !closed.get();
+    }
+
+    /**
+     * Sets whether background cleanup is suspended or not.
+     *
+     * @param value Boolean indicating whether to suspend background processing or not.
+     */
+    void setSuspended(boolean value) {
+        suspended.set(value);
+    }
+
+    /**
+     * Adds given chunks to list of garbage chunks.
+     *
+     * @param chunksToDelete List of chunks to delete.
+     */
+    void addToGarbage(Collection<String> chunksToDelete) {
+        val currentTime = currentTimeSupplier.get();
+
+        chunksToDelete.forEach(chunkToDelete -> addToGarbage(chunkToDelete, currentTime + config.getGarbageCollectionDelay().toMillis(), 0));
+
+        if (queueSize.get() >= config.getGarbageCollectionMaxQueueSize()) {
+            log.warn("{}: deleteGarbage - Queue full. Could not delete garbage. Chunks skipped", traceObjectId);
+        }
+    }
+
+    /**
+     * Adds given chunk to list of garbage chunks.
+     *
+     * @param chunkToDelete Name of the chunk to delete.
+     * @param startTime Start time.
+     * @param attempts Number of attempts to delete this chunk so far.
+     */
+    void addToGarbage(String chunkToDelete, long startTime, int attempts) {
+        if (queueSize.get() < config.getGarbageCollectionMaxQueueSize()) {
+            garbageChunks.add(new GarbageChunkInfo(chunkToDelete, startTime, attempts));
+            queueSize.incrementAndGet();
+        } else {
+            log.debug("{}: deleteGarbage - Queue full. Could not delete garbage. chunk {}.", traceObjectId, chunkToDelete);
+        }
+    }
+
+    /**
+     * Delete the garbage chunks.
+     *
+     * This method retrieves a few eligible chunks for deletion at a time.
+     * The chunk is deleted only if the metadata for it does not exist or is marked inactive.
+     * If there are any errors then failed chunk is enqueued back.
```

Comment:
> If there are any errors then failed chunk is enqueued back.

To describe the actual behavior, we would need to add `up to a max number of attempts`, right?